### PR TITLE
[MOBILE-801] new BLoC-based sample

### PR DESCRIPTION
**Pull request by crow**
### What do these changes do?
Adds a new sample that implements the primary SDK features. Implements BLoC architecture.

### Why are these changes necessary?

To have a sample that implements core Urban Airship features.

### How did you verify these changes?

Manual testing.

#### Verification Screenshots:

https://slack-files.com/T025Q1VP7-FMJS48KHD-1da34c3b2c

### CODE REVIEW

File: android/.idea/vcs.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (BrianBatchelder) [inline android/.idea/vcs.xml]: Add this directory to .gitignore

File: example/lib/bloc/airship_bloc.dart [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+  final _messagesStreamController = StreamController<List<InboxMessage>>.broadcast();
+  final _messageAddedStreamController = StreamController<InboxMessage>();
+  final _messageRemovedStreamController = StreamController<InboxMessage>();
+
+  Stream<String> get channelStream => _channelStreamController.stream;
+  StreamSink<String> get channelSink => _channelStreamController.sink;
+
+  Stream<String> get namedUserStream => _namedUserStreamController.stream;
+  StreamSink<String> get namedUserSink=> _namedUserStreamController.sink;
```
Comment (BrianBatchelder) [inline example/lib/bloc/airship_bloc.dart]: whitespace

File: example/lib/bloc/airship_bloc.dart [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      });
+    });
+
+    Airship.userNotificationsEnabled.then((val) {
+      _notificationsEnabledStreamController.add(val);
+    });
+
+    _notificationsEnabledSetStreamController.stream.listen((enabled) {
+      Airship.setUserNotificationsEnabled(enabled).then((val) {
+        notificationsEnabledSink.add(enabled);
```
Comment (BrianBatchelder) [inline example/lib/bloc/airship_bloc.dart]: should `enabled` be `val` in case `setUserNotificationsEnabled` wasn't able to do it?

File: example/lib/bloc/airship_bloc.dart [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+      debugPrint('Inbox updated link');
+      // Get fresh message list
+      Airship.inboxMessages.then((messages) {
+        messagesSink.add(messages);
+      });
+    });
+
+    Airship.onShowInboxMessage.listen((messageId) {
+      //MaterialPageRoute(builder: (context) => MessageView(message: message));
+      debugPrint('Show inbox message $messageId');
```
Comment (BrianBatchelder) [inline example/lib/bloc/airship_bloc.dart]: Is the actual `show` implemented somewhere else?

File: example/lib/bloc/airship_bloc.dart [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+      });
+    });
+
+    Airship.onChannelRegistration.listen((event) {
+      _channelUpdatedStreamController.add(event.channelId);
+      debugPrint('Channel registration $event');
+    });
+
+    Airship.onPushReceived.listen((event) {
+      debugPrint('Push Received $event');
```
Comment (BrianBatchelder) [inline example/lib/bloc/airship_bloc.dart]: Should there be a stream for `onPushReceived`? Does the app just listen to `Airship.onPushReceived` elsewhere? These questions apply to the next three events also.

File: example/lib/widgets/text_add_bar.dart [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+                    border: InputBorder.none,
+                    hintText: label
+                ),
+                //tyle: Styles,
+                cursorColor: Styles.airshipBlue,
+              ),
+            ),
+            GestureDetector(
+              onTap: () {
+                SystemChannels.textInput.invokeMethod('TextInput.hide');
```
Comment (BrianBatchelder) [inline example/lib/widgets/text_add_bar.dart]: weird. surprised there isn't a `SystemChannels.textInput.hide()` method

Comment (crow) [reply example/lib/widgets/text_add_bar.dart]: not sure why this is, but this works well for now

File: example/lib/bloc/airship_bloc.dart (hunk 5 above)
Comment (crow) [reply example/lib/bloc/airship_bloc.dart]: Going to flesh these out in this follow up ticket:
https://urbanairship.atlassian.net/browse/MOBILE-849

I think this greatly improves on what we have. Adding deep linking is going to add some complexity and I think it's best to tackle it in a separate PR so it doesn't become impossible to review.

File: example/lib/bloc/airship_bloc.dart (hunk 4 above)
Comment (crow) [reply example/lib/bloc/airship_bloc.dart]: Same as below, this is going to require the work to support deep linking, which requires some research on my part - and will be best to review as a separate PR.

File: example/lib/bloc/airship_bloc.dart (hunk 3 above)
Comment (crow) [reply example/lib/bloc/airship_bloc.dart]: fixed

File: example/lib/bloc/airship_bloc.dart [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+  AirshipBloc() {
+    Airship.channelId.then((val) {
+      _channelStreamController.add(val);
+    });
+
+    Airship.tags.then((val) {
+      _tagsStreamController.add(val);
+    });
+
+    _tagsAddedStreamController.stream.listen((tags) {
```
[the remark this answers is not in the document]
Comment (crow) [reply example/lib/bloc/airship_bloc.dart]: messagesSteam is public, tagsAddedSteam is not. Streams that are accessed outside of this class are given a blahStream property, otherwise it's fine to access the streamController directly to get at its stream property.

File: example/lib/screens/home.dart [hunk 8]
```diff
@@ -0,0 +1,83 @@
+import 'package:flutter/material.dart';
+import 'package:airship_example/styles.dart';
+import 'package:airship_example/bloc/airship_bloc.dart';
+
+class _HomeState extends State<Home> {
+  final AirshipBloc _airshipBloc = AirshipBloc();
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
```
Comment (marc-scig) [inline example/lib/screens/home.dart]: is it possible to break this into smaller pieces and combine them? I find it pretty hard to follow with all the nesting.

File: example/lib/screens/message_center.dart [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+import 'package:airship_example/screens/message_view.dart';
+import 'package:airship_example/bloc/bloc.dart';
+
+class MessageCenter extends StatelessWidget {
+  final AirshipBloc _airshipBloc = AirshipBloc();
+
+  @override
+  Widget build(BuildContext context) {
+    Widget _buildMessageList(List<InboxMessage> messages) {
+      //messages = List<InboxMessage>.from(messages);
```
Comment (marc-scig) [inline example/lib/screens/message_center.dart]: remove?

File: example/lib/screens/message_view.dart [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    );
+  }
+
+  void onInboxMessageViewCreated(InboxMessageViewController controller) {
+    Airship.inboxMessages.then((List<InboxMessage> messages) {
+      if (message != null) {
+        debugPrint("Loading message: ${message.title}");
+        controller.loadMessage(message);
+      } else {
+        debugPrint("Error message: $messages");
```
Comment (marc-scig) [inline example/lib/screens/message_view.dart]: what are we indicating here, that the message was unexpectedly null?

File: example/lib/screens/settings.dart [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        ),
+        backgroundColor: Colors.white,
+        body: ListView(
+            children: ListTile.divideTiles(
+              context: context,
+              tiles: [
+                StreamBuilder(
+                    stream: _airshipBloc.notificationsEnabledStream,
+                    builder: (context, snapshot) { return SwitchListTile(
+                      //leading:Icon(Icons.settings_input_antenna),
```
Comment (marc-scig) [inline example/lib/screens/settings.dart]: removable?

File: example/lib/bloc/airship_bloc.dart [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    Airship.onInboxUpdated.listen((event) {
+      debugPrint('Inbox updated link');
+      // Get fresh message list
+      Airship.inboxMessages.then((messages) {
+        messagesSink.add(messages);
+      });
+    });
+
+    Airship.onShowInboxMessage.listen((messageId) {
+      //MaterialPageRoute(builder: (context) => MessageView(message: message));
```
[the remark this answers is not in the document]
Comment (crow) [reply example/lib/bloc/airship_bloc.dart]: This is a stub for implementing deep linking. I wanted to separate out that work because I think it's going to add a reasonable amount of complexity.

[the remark this answers is not in the document]
Comment (crow) [reply example/lib/bloc/airship_bloc.dart]: I can remove it

File: example/lib/screens/home.dart [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+                              child: Text( "Enable Push",
+                                style: Styles.homeButtonText,
+                              ),
+                              color: Styles.airshipRed,
+                              shape: StadiumBorder(),
+                              height: 40,
+                              minWidth: 400,
+                              padding: EdgeInsets.symmetric(vertical: 35),
+                              onPressed: () {
+                                // TODO
```
[the remark this answers is not in the document]
Comment (crow) [reply example/lib/screens/home.dart]: removed

File: example/lib/bloc/airship_bloc.dart [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+  final _tagsStreamController = StreamController<List<String>>.broadcast();
+  final _tagsAddedStreamController = StreamController<List<String>>();
+  final _tagsRemovedStreamController = StreamController<List<String>>();
+
+  final _messagesStreamController = StreamController<List<InboxMessage>>.broadcast();
+  final _messageAddedStreamController = StreamController<InboxMessage>();
+  final _messageRemovedStreamController = StreamController<InboxMessage>();
+
+  Stream<String> get channelStream => _channelStreamController.stream;
+  StreamSink<String> get channelSink => _channelStreamController.sink;
```
[the remark this answers is not in the document]
Comment (crow) [reply example/lib/bloc/airship_bloc.dart]: removed

File: example/lib/bloc/airship_bloc.dart [hunk 15]
```diff
@@ -0,0 +1,154 @@
+import 'dart:async';
+import 'package:airship_example/bloc/bloc_base.dart';
+import 'package:airship/airship.dart';
+import 'package:flutter/material.dart';
+
+class AirshipBloc implements BlocBase {
+  final _channelStreamController = StreamController<String>();
+  final _channelUpdatedStreamController = StreamController<String>();
```
[the remark this answers is not in the document]
Comment (crow) [reply example/lib/bloc/airship_bloc.dart]: nah just leftover cruft

File: example/lib/screens/settings.dart (hunk 11 above)
Comment (crow) [reply example/lib/screens/settings.dart]: removed

File: example/lib/screens/message_view.dart (hunk 10 above)
Comment (crow) [reply example/lib/screens/message_view.dart]: should just print "Attempted to load message view for null message", fixed.

File: example/lib/screens/home.dart (hunk 8 above)
Comment (crow) [reply example/lib/screens/home.dart]: removed the button into its own widget to make this a little simpler

File: example/lib/bloc/airship_bloc.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (BrianBatchelder) [reply example/lib/bloc/airship_bloc.dart]: OK
